File: apple/DemoResources/Managers/PlaybackManager.m
//
//  PlaybackManager.m
//  SurgeiOSDemoApp
//
//  Created by Niall Kelly on 10/10/2017.
//  Copyright © 2017 Instil. All rights reserved.
//

#import "PlaybackManager.h"
#import "PlaybackStream.h"
#import "RtspAddress.h"

NSString *const PlaybackStreamsDidChangeNotification = @"PlaybackStreamsDidChangeNotification";

@interface PlaybackManager()
@property (strong, nonatomic) NSMutableArray<PlaybackStream *> *streams;
@end

@implementation PlaybackManager

- (instancetype)init
{
  self = [super init];
  if (self) {
    _streams = [[NSMutableArray<PlaybackStream *> alloc] initWithCapacity:0];
  }
  return self;
}

- (void)startStreamWithAddress:(RtspAddress *)address {
  NSUInteger idx = [self.streams indexOfObjectPassingTest:^BOOL(PlaybackStream * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
    return [obj.rtspAddress.address isEqualToString:address.address];
  }];
  if (idx < self.streams.count) return;
  PlaybackStream *stream = [[PlaybackStream alloc] initWithRtspAddress:address];
  [stream start];
  [_streams addObject:stream];
  [[NSNotificationCenter defaultCenter] postNotificationName:PlaybackStreamsDidChangeNotification object:nil];
}

- (void)removeStream:(PlaybackStream *)stream {
  [stream stop];
  [_streams removeObject:stream];
  [[NSNotificationCenter defaultCenter] postNotificationName:PlaybackStreamsDidChangeNotification object:nil];
}

- (void)removeStreamAtIndex:(NSUInteger)idx {
  if (idx >= self.streams.count) return;
  [self removeStream:self.streams[idx]];
}

@end

File: apple/SurgeiOSSwiftDemo/SurgeiOSSwiftDemo/PlaybackManager.swift
//
//  PlaybackManager.swift
//  SurgeiOSSwiftDemo
//
//  Created by Stevey Brown on 21/08/2018.
//  Copyright © 2018 Instil. All rights reserved.
//

import Foundation
import SurgeiOS

class PlaybackManager: NSObject {
    private(set) var streams: [PlaybackStream] = []
    
    // MARK: API
    
    func startStreamWithAddress(_ address: RtspStream) {
        let stream = PlaybackStream(withRSTPStream: address)
        stream.start()
        self.streams.append(stream)
    }
    
    func removeStream(_ stream: PlaybackStream) {
        stream.stop()
        if let index = self.streams.index(of: stream) {
            self.streams[index].player.playerView = UIImageView()
            self.streams.remove(at: index)
        }
    }
}
